lib/src/license_detection/confidence.dart

```diff
@@ -0,0 +1,260 @@
+// Copyright (c) 2021, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+part of 'license_detector.dart';
+
+/// An instance indicating a mismatch between unknown text and a known license.
+@sealed
+@visibleForTesting
+class LicenseMismatchException implements Exception {
+  final String message;
+
+  LicenseMismatchException(this.message);
+
+  @override
+  String toString() {
+    return message;
+  }
+}
+
+/// Computes the confidence of [knownLicense] matching with [unknownLicense] and
+/// returns an instance of [LicenseMatch].
+///
+/// Throws if an unaccepatable change was made or the calculated confidence is
+/// lesser than the set threshold.
+@visibleForTesting
+LicenseMatch licenseMatch(
+  LicenseWithNGrams unknownLicense,
+  LicenseWithNGrams knownLicense,
+  MatchRange matchRange,
+  double threshold,
+) {
+  if (unknownLicense.n != knownLicense.n) {
+    throw LicenseMismatchException(
+        "Can't comapare the licenses due to different granularity");
+  }
+  final diffs = getDiffs(
+    unknownLicense.tokens,
+    knownLicense.tokens,
+    matchRange,
+  );
+
+  final range = diffRange(tokensNormalizedValue(knownLicense.tokens), diffs);
+  final valuationDiffs = diffs.skip(range.start).take(range.end - range.start);
+
+  // Make an initial check on the diffs to see if their are any
+  // unacceptable substitutions made.
+  verifyNoVersionChange(valuationDiffs, knownLicense.identifier);
+  verifyNoGplChange(valuationDiffs, knownLicense.identifier);
+
+  final distance = scoreDiffs(valuationDiffs);
+  final confidence = confidencePercentage(
+    knownLicense.tokens.length,
+    distance,
+  );
+
+  if (confidence < threshold) {
+    throw LicenseMismatchException(
+        'Confidence $confidence is lesser than threshold $threshold');
+  }
+
+  return LicenseMatch(
+    unknownLicense.tokens
+        .skip(matchRange.input.start)
+        .take(matchRange.input.end - matchRange.input.start)
+        .toList(),
+    confidence,
+    knownLicense,
+    diffs,
+    Range(range.start, range.end),
+  );
+}
+
+/// Calculates the confidence percentage as 1 subtracted by the ratio of
+/// Levenshtein word distance to the number of tokens in known license.
+@visibleForTesting
+double confidencePercentage(int knownLength, int distance) {
+  if (knownLength == 0) {
+    return 1.0;
+  }
+  final confidence = 1.0 - (distance / knownLength);
```

Comment:
The distance can be larger than length if the number of additions or deletions of words is greater than the known length and they should be different if that is the case it will never reach this stage as it will be filtered out at `primary filter` stage and also the `find potential matches` will not pick up as the there should be at least `theshold * known length` tokens must be present to qualify as a match potential match and reach to this stage of scoring.

Reply:
Let's add `assert(distance <= knownLength);` in case the other stages change in the future.

Reply:
Done